public/app/plugins/datasource/prometheus/datasource.ts

```diff
@@ -645,106 +656,126 @@ export class PrometheusDatasource extends DataSourceWithBackend<PromQuery, PromO
-    const eventList: AnnotationEvent[] = [];
-    const splitKeys = tagKeys.split(',');
+    return await lastValueFrom(
+      getBackendSrv()
+        .fetch<BackendDataSourceResponse>({
+          url: '/api/ds/query',
+          method: 'POST',
+          data: {
+            from: (this.getPrometheusTime(options.range.from, false) * 1000).toString(),
+            to: (this.getPrometheusTime(options.range.to, true) * 1000).toString(),
+            queries: [queryModel],
+          },
+          requestId: `prom-query-${annotation.name}`,
+        })
+        .pipe(
+          map((rsp: FetchResponse<BackendDataSourceResponse>) => {
+            return this.processsAnnotationResponse(options, rsp.data);
+          })
+        )
+    );
+  }
 
-    if (isFetchErrorResponse(response) && response.cancelled) {
+  processsAnnotationResponse = (options: any, data: BackendDataSourceResponse) => {
+    const frames: DataFrame[] = toDataQueryResponse({ data: data }).data;
+    if (!frames || !frames.length) {
       return [];
     }
 
-    const step = Math.floor(query.step ?? 15) * 1000;
-
-    response?.data?.data?.result?.forEach((series) => {
-      const tags = Object.entries(series.metric)
-        .filter(([k]) => splitKeys.includes(k))
-        .map(([_k, v]: [string, string]) => v);
-
-      series.values.forEach((value: any[]) => {
-        let timestampValue;
-        // rewrite timeseries to a common format
-        if (annotation.useValueForTime) {
-          timestampValue = Math.floor(parseFloat(value[1]));
-          value[1] = 1;
-        } else {
-          timestampValue = Math.floor(parseFloat(value[0])) * 1000;
-        }
-        value[0] = timestampValue;
-      });
-
-      const activeValues = series.values.filter((value) => parseFloat(value[1]) >= 1);
-      const activeValuesTimestamps = activeValues.map((value) => value[0]);
+    const annotation = options.annotation;
+    const { tagKeys = '', titleFormat = '', textFormat = '' } = annotation;
+
+    const step = rangeUtil.intervalToSeconds(annotation.step || ANNOTATION_QUERY_STEP_DEFAULT) * 1000;
+    const tagKeysArray = tagKeys.split(',');
+    const frame = frames[0];
+    const timeField = frame.fields[0];
+    const valueField = frame.fields[1];
+    const labels = valueField?.labels || {};
+
+    const tags = Object.keys(labels)
+      .filter((label) => tagKeysArray.includes(label))
+      .map((label) => labels[label]);
+
+    const timeValueTuple: Array<[number, number]> = [];
+
+    let idx = 0;
+    valueField.values.forEach((value) => {
+      let timeStampValue: number;
+      let valueValue: number;
+      const time = timeField.values.get(idx);
+
+      // If we want to use value as a time, we use value as timeStampValue and valueValue will be 1
+      if (options.annotation.useValueForTime) {
+        timeStampValue = Math.floor(parseFloat(value));
+        valueValue = 1;
+      } else {
+        timeStampValue = Math.floor(parseFloat(time));
```

Comment:
Why is `* 1000` not needed anymore?

Reply:
Because when we were getting results directly from prometheus, we were getting the [timestamp in seconds](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries). However, now we process it to dataFrame on backend and we use timestamp in milliseconds, as Grafana uses milliseconds.